### ISSUE TITLE
Consider advanced integration tests

### ISSUE THREAD

**Issue by TTWNO** [odilia-app/atspi#127]
I'm unsure how well it would fit into the `cargo` build system, but it would be nice if we could build a few `Qt`, `Gtk`, and maybe `egui` apps to get an idea of more end-to-end testing.

Or would this be the responsibility of the individual toolkits?

**Comment by TTWNO** [comment]
Want to revive this idea:

I wonder if we can have a bunch of example implementations that (if all of them are run), would produce at least one of every (non-deprecated) event.

**Comment by TTWNO** [comment]
Then, we can reference which examples actually make those events happen in the docs for them.

**Comment by TTWNO** [comment]
Much has been said on Matrix about this, but I want to write down some of the thoughts that we have had over the years:

## Application-Level A11y Checks (for consumers)

Test real applications (GTK, Qt, etc.) this is good and would be a first step in making sure that _consumer_ accessibility is tested with _real world_ systems. It also gives us a chance to see where we might deviate from the step (or the toolkits).

## Generic `git` testing.

Upstream sometimes get out of sync with XML (or its associated comments). I'm wondering if we could build a tester binary/program that can be added as a Github/Gitlab action (or maybe work with `at-spi2-core`) to ensure compliance with a series of test applications. With test coverage messured in how many interfaces/methods/properties function correctly. (Correctly counts as anything which does not error or crash under acceptable parameters).

[The recent case where many applications give the wrong parent for `/org/a11y/atspi/accessible/root`](https://github.com/odilia-app/atspi/issues/271#issuecomment-2997361107) is a good example of something relatively trivial that would be much better to catch earlier rather than later.

I think we could pretty reasonably write a "gauntlet" application that just _queries everything it can_ and report back on results.

Here are some baseline ideas:

1. `/org/a11y/atspi/accessible/root` _must_ exist, implement Application, and its parent must be ("", "/org/a11y/atspi/null").
2. All interfaces/fields that any object implements _must work_: no errors, no segfaults, no crashing in general.
3. If implemented, `org.a11y.atspi.Cache`'s `GetItems` should return consistent values compared to calling the methods/getting the properties one-by-one.
4. Given invalid inputs (like an index that's too big) all applications should _error_, not crash, not segfault, not throw garbage into the bus. We should be able to read an error back from the message—we still _must_ get a response.

That would at least be a good start. It would catch things like [the wrong method names being used](https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8652), [incorrect signatures of events](https://github.com/odilia-app/atspi/issues/220), and [object ref edge cases](https://github.com/odilia-app/atspi/issues/271) in CI instead of finding these things out on a bi-monthly basis. 

If their pipelines fail, with easily understandable error messages, hopefully this will help push a11y to be more consistent.

cc @luukvanderduim thoughts?

I think a lot of what you and Colton have been doing making lots of examples, tracking down various properties of the running applications, etc. is a very good introduction to how this might work in practice. Just with more interfaces/mmethods/property getters and _many_ more assertions.

**Comment by luukvanderduim** [comment]
I think it is a great idea to have a crate with one or many test applications and/or ci tests in a crate that help us more realistically test atspi's surfaces.

A separate crate, but included in actions is perfect.

Downstream users usually do not appreciate it if we add dependencies, even if these are for testing purposes only, be it toolkits or executors, they all add to our greater dependency tree `Cargo.lock` - so a separate crate is perfect, there we are not sensitive about growing dependency trees.

@C-Loftus What do you think?

**Comment by TTWNO** [comment]
Don't mind me, just writing down ideas so I don't lose them. Also, they're all disconnected, don't worry about me meandering, haha!

We can try building a tree from the application, and if it fails part way through, we could probably error out with the partial tree as output. Optionally, we may want to consider rich output for these types, since some of those trees get pretty big and might be hard to parse through (also, printing line and box-drawing characters are usually inaccessible).

While building the tree, we can test additional properties on each item: roles/text interfaces (if available)/action interface (if available)/etc. These tests would need to pass for every item we come across.

I also thought about maybe trying to abuse `llvm-cov` debug info so that we can report coverage of the XML files? Could be dumb, just an idea.

We could also use XLST to style the XML files based on reporting data?

[I remember an article](https://fasterthanli.me/articles/aiming-for-correctness-with-types) where Mr. FasterThanLime was talking about running a test-suite on his own crate, and that he did some kind of macro magic to automatically test new functionality.

Now here are some questions that maybe others can answer:

- How do we group tests? By application, by interface, by object? To me, the most obvious way is Applications -> Interfaces -> Methods -> Objects. But this makes it hard to deal with tests where multiple errors happen on one object... there must be some kind of format to deal with this?
- And in general, I'm looking for tools in this area that _already exist_, as much as possible, I do not want to be making new formats/views/HTML output mechanisms. If there is any standard way to do stuff like this, we should use it.

**Comment by C-Loftus** [comment]
> [@C-Loftus](https://github.com/C-Loftus) What do you think?

All makes sense to me! Two comments that are less important:

- If tests with non-trivial setup are added, I wonder if it makes sense to have a simple task runner (makefile, pre-commit.yml, justfile, something else) for clearly showing a user how to run them
 
- Once a baseline is implemented, it would be very interesting to be able to test against not only different toolkits but also the same app on different desktop environments or packaging formats (flatpak, snap etc) to catch stuff like #257 . 
    - I wonder if there is any world in which we can spin up ephemeral Desktop Linux test containers. That would obviously be very heavy though, and not suitable for ci. Moreso a long term ideal.

**Comment by TTWNO** [comment]
> I wonder if there is any world in which we can spin up ephemeral Desktop Linux test containers. That would obviously be very heavy though, and not suitable for ci. Moreso a long term ideal.

Not a problem if we make our own images. It's only difficult if we need to recreate the steps every time.
